### DISCUSSION TITLE
[Support] 请问bull如何控制任务执行的频率？

### DISCUSSION THREAD

**Discussion by phoooob** [midwayjs/midway#3591] [Q&A] [answered]
### Describe the problem（描述问题）

如果我有3台服务器，每台服务器启动4个实例，我想控制每台服务器在1秒内最多同时执行2个任务，应该如何设置？我是这么写的，但是却不生效，同一时间依然会有4个任务在执行。
```
import { IProcessor, Processor } from '@midwayjs/bull';
import { Inject } from '@midwayjs/decorator';
import { ILogger } from '@midwayjs/logger';
import { sleep } from '@midwayjs/bootstrap/dist/util';

@Processor(
  'test',
  {
    removeOnComplete: false,
    removeOnFail: false,
    attempts: 99999999,
    backoff: {
      type: 'fixed',
      delay: 10000,
    },
  },
  {
    limiter: {
      max: 2,
      duration: 1000,
      // groupKey: getServerIP(),
      // bounceBack: false,
    },
  }
)
export class PdfTask implements IProcessor {
  @Inject() private logger: ILogger;

  async execute() {
    console.log('ID:25，execute 任务开始', new Date().toLocaleString());
   
    await sleep(5000);
    console.log('ID:29，execute 任务结束', new Date().toLocaleString());
   
  }
}

```

### Midway Versions（Midway 版本）

"@midwayjs/bull": "^3.11.15",
 "@midwayjs/core": "^3.11.11",

**Comment by czy88840616** [comment]
启动两个进程即可，现在默认应该是每个进程执行一个 job。

**Reply by phoooob** [reply]
希望能充分利用多核服务器资源，启动4个实例。bull没有现成的解决方案吗？

**Reply by czy88840616** [reply]
启动多个midway进程，每个进程执行一个bull任务，这就是利用了多核呀

**Reply by phoooob** [reply]
是我没说清楚。
我的意思是我们的服务器是8核的，所以希望能最少启动4个应用实例，以便最大程度利用cpu资源，提升并发处理能力，这个完全是应用业务侧的需求。
bull管理的队列只是项目中的一个业务，不能因为这一个业务，把4个应用实例减少为2个。
但是这个队列的任务执行时候特别消耗cpu、内存，希望每台服务器上同一时间最多只有2个实例在执行队列中的任务，而不是4个实例都在执行这个队列的任务。

**Reply by czy88840616** [reply]
你的需求其实可以这么分解，一部分执行普通任务（http 服务，4c），一部分提供 bull 服务（计划调度，2c），这种资源不均的场景下，最合理的是启动不同的应用，一个应用跑 4 个进程，一个应用跑 2 个进程，当然这两个应用的代码是可以复用的（一个仓库项目）

**Reply by phoooob** [reply]
因为这个队列的任务实时性要求不高，另外我们的业务会集中到某个时间段繁忙，繁忙期过了就空闲了。所以希望能利用服务器空闲的时候处理队列任务，同时不要把服务器搞崩溃掉。暂时不考虑让这个队列的业务独占一台服务器的方案。
bull 没有可以满足我的需求的方案吗？在哪里可以看到 @midwayjs/bull 的源码啊？

**Reply by czy88840616** [reply]
不是让你单独部署服务器，而是在同一台服务器上部署不同的入口。https://github.com/midwayjs/midway/tree/main/packages/bull

**Reply by phoooob** [reply]
任务是由普通的http服务下发的，如果分两个应用部署的话，那没办法下发任务了啊

**Reply by czy88840616** [reply, marked as answer]
任务是由 bull（redis）下发，和进程或者应用无关，完全可以一个 http 服务添加任务，另外的 bull 服务接收任务信息执行。

**Reply by phoooob** [reply]
明白了，不使用@Processor装饰器来创建消费者，而是通过queue.process() 来手动添加消费者。已解决，感谢！
